js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,171 @@
+import * as Mithril from 'mithril';
+
+import Model from '../Model';
+
+export interface Page<TModel> {
+  number: number;
+  items: TModel[];
+
+  // // maybe we can put page links here?
+  // links: {[name: string]: string},
+  // // or
+  // previousPage?: string;
+  // nextPage?: string;
+}
+
+export interface PaginationLocation {
+  page: number;
+  startIndex?: number;
+  endIndex?: number;
+}
+
+export default abstract class PaginatedListState<T extends Model> {
+  // TODO:
+  //  - what does this page track? last page request? last page that is loaded
+  //      (e.g. pg 7 if u have 3-7 loaded)
+  //  - When do we set it?
+  protected location!: PaginationLocation;
+
+  // TODO: might want interface so pagination can work w/o model-specific stuff
+  protected type: string;
+  protected offset: number;
+
+  protected pages: Page<T>[] = [];
+  protected params: any = {};
+
+  protected initialLoading: boolean = false;
+  protected loadingPrev: boolean = false;
+  protected loadingNext: boolean = false;
+
+  hasPrev: boolean = false;
+  hasNext: boolean = false;
+
+  // TODO: this constructor is for class that implements
+  // what data do we want passed?
+  protected constructor(type: string, offset: number = 20) {
```

Comment:
If this is an abstract class, shouldn't type be static?

We could also do an initial page of data, so that if there's stuff in the preloaded API doc we can reuse that without immediately triggering an API call.

Reply:
> shouldn't type be static?

That makes more sense, yes. 

> We could also do an initial page of data, so that if there's stuff in the preloaded API doc we can reuse that without immediately triggering an API call.

Hm this might work better as another method that is called from the subclass constructor 🤔

Reply:
Static properties that are overriden in subclasses is a bit more scuffed in JS than PHP.

Do we *need* a static class or do we want to use an abstract getter, for example (wouldn't be static)
![image](https://user-images.githubusercontent.com/6401250/115636365-cdd8b100-a2db-11eb-83de-4d34628adb20.png)

Reply:
Dont think it matters that much, abstract method or getter is probs fine.

Reply:
Currently got is as static. Will probs change to getter then (will also make TS error disappear)